Extract transport agnostic zipkin code into finagle-zipkin-core

sveinnfannar opened the pull request:
Problem

Address issue https://github.com/twitter/finagle/issues/465 (Add http and kafka transports to zipkin-tracer).
Zipkin can receive traces via Kafka, HTTP and Scribe. Some Finagle users (including Quizup) already using Kafka are reluctant to introduce Scribe into their stack as it is no longer maintained.
So the purpose of this PR is to support Kafka as a transport for Zipkin traces, as well as structuring the code such that it allows other transports to be added (like HTTP).


Solution

Extract most of finagle-zipkin into finagle-zipkin-core, leaving finagle-zipkin with only scribe specific stuff (with no api breaking changes to finagle-zipkin of course). ~~Then implement finagle-zipkin-kafka, which depends on finagle-zipkin-core~~. Kafka transport will be introduced in a separate PR.

This structure was initially suggested by @kevinoliver in https://github.com/twitter/finagle/issues/465#issuecomment-183012155. 


Todo

 - [x] Extract common code into finagle-zipkin-core (largest part)
 - [ ] ~~Implement finagle-zipkin-kafka (in progress)~~ (separate PR)
 - [x] Clean up Pants and SBT dependencies


Questions

@mosesn can you check this out and provide some feedback?
@adriancole is this in line with what you had envisioned?

adriancole commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/SamplingTracer.scala:
```diff
@@ -0,0 +1,160 @@
+package com.twitter.finagle.zipkin.core
```
weird this didn't get caught as a rename..

sveinnfannar replied:
yes, it did in my local git diff

sveinnfannar replied:
btw, I fixed up the commits so this appears as a rename

adriancole commented on finagle-zipkin-core/src/main/scala/BUILD:
```diff
@@ -0,0 +1,21 @@
+    '3rdparty/jvm/com/fasterxml/jackson/core:jackson-core',
+    '3rdparty/jvm/com/fasterxml/jackson/core:jackson-databind',
+    '3rdparty/jvm/com/fasterxml/jackson/module:jackson-module-scala',
+    '3rdparty/jvm/org/apache/thrift:thrift-0.5.0',
```
one option is to drop this dep and instead use io.zipkin.java:zipkin:1.0.0 as that includes an implicit codec. In other words change to use the canonical model library.

sveinnfannar replied:
Very good point. Although I'm going to try keeping the scope of these changes as small as possible. Let's look into it when the rest is done.

adriancole commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala:
```diff
@@ -0,0 +1,149 @@
+    annotationType: thrift.AnnotationType
+  ) {
+    spanMap.update(record.traceId) { span =>
+      span.addBinaryAnnotation(BinaryAnnotation(key, value, annotationType, span.endpoint))
```
ex in my local workspace, I made this a map of zipkin.Span.Builder.. not required, as such a change is internal-only and could be layered as easily later

adriancole commented:
added a couple notes, none of which are important :) carry on. looks like the right path

adriancole commented on finagle-zipkin/src/main/scala/com/twitter/finagle/zipkin/thrift/ScribeRawZipkinTracer.scala:
```diff
@@ -0,0 +1,184 @@
+import scala.collection.mutable.ArrayBuffer
+import scala.language.reflectiveCalls
+
+object ScribeRawZipkinTracer {
```
Is there a META-INF change needed? or is the service loader unaffected by the rename

sveinnfannar replied:
Yes it should be unaffected.
`META-INF/services/com.twitter.finagle.tracing.Tracer` in finagle-zipkin points to [`com.twitter.finagle.zipkin.thrift.SamplingTracer`](https://github.com/twitter/finagle/pull/513/files#diff-06cb96e4d4229bad23c2e59316178504R61).

sveinnfannar commented:
This diff is too large, I'm going to do some git foo to make it easier to review.

sveinnfannar commented:
Ok I force pushed a change that separates the moving of files from the actual code changes. The PR is now easier to review commit by commit.

adriancole commented on finagle-zipkin-core/src/test/scala/com/twitter/finagle/zipkin/core/RawZipkinTracerTest.scala:
```diff
@@ -17,55 +16,16 @@ class RawZipkinTracerTest extends FunSuite {
-      response
+  class FakeRawZipkinTracer extends RawZipkinTracer(NullStatsReceiver) {
+    var spans: Seq[Span] = Seq.empty
+    override def sendSpans(xs: Seq[Span]): Future[Unit] = {
```
cool basically the extender just needs to implement sendSpans

mosesn commented:
This looks good to me.  @sveinnfannar could you add a note to CHANGES?

@sveinnfannar what do you think about splitting this up into a few PRs? when we merge it internally, it's all going to be squashed, and I think it would be nice to have separate commits for each piece.  ie, the first PR would be HEAD => First Milestone, second commit would be HEAD => First Milestone => Second Milestone, etc, and for the first patch I would grab HEAD => First Milestone, second patch I would grab First Milestone => Second Milestone, etc.

I would say this first milestone is pretty close to being ready to ship.

I would also suggest breaking out the Scribe stuff to a separate module at some point in time if possible, but it's much less necessary than the work you've outlined so far.

sveinnfannar commented:
Thanks for the review @mosesn.

Ok, I'll rename this PR and put the finishes touches on it.

Regarding breaking out the scribe stuff, do you mean something like having a module named `finagle-zipkin-scribe`?
`finagle-zipkin` already contains only scribe specific stuff. We could rename it to `finagle-zipkin-scribe`. Then for backwards compatibility we would have a `finagle-zipkin` module that just depends on `finagle-zipkin-scribe`.
It would basically only contain a service loader and something like this:
```scala
package com.twitter.finagle.zipkin.thrift
@deprecated("Use com.twitter.finagle.zipkin.scribe.ScribeZipkinTracer instead", "6.36.0")
class ZipkinTracer(..) extends ScribeZipkinTracer(..)
```

mosesn commented:
Hmm, I thought the service-loading bit was in finagle-zipkin, but if it's somewhere else then I see what you mean.  Should actually be fine this way, imo.

adriancole commented:
+1 merging the structural changes first makes it an easy merge, and paves
the way for multiple transports being reviewed or added independently (ex
http, kafka)

sveinnfannar commented:
@mosesn I finished some minor tweaks and manual testing so this PR should be ready. The notes in CHANGES still need an RB_ID. Is that something you add internally?

kevinoliver commented:
@sveinnfannar yeah we'll add the RB_ID, thanks. i'll try to take a look today as well.

kevinoliver commented on CHANGES:
```diff
@@ -56,6 +56,17 @@ Breaking API Changes
   * finagle-memcached: `c.t.f.memcached.Client` now uses `c.t.bijection.Bijection`
     instead of `c.t.u.Bijection`. ``RB_ID=834383``
 
+<<<<<<< HEAD
```
fyi merge conflict?

sveinnfannar replied:
:$

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/Annotation.scala:
```diff
@@ -1,4 +1,4 @@
-package com.twitter.finagle.zipkin.thrift
+package com.twitter.finagle.zipkin.core
```
moving the package is a breaking change. i think its probably worth it, but please note these in the "Breaking" section of `CHANGES`.

sveinnfannar replied:
In the breaking section I listed the public classes (and companion objects) affected.:

>   * finagle-zipkin: Moved case classes `Span`, `ZipkinAnnotation`,
>     `BinaryAnnotation`, `Endpoint` and `Sampler` to finagle-zipkin-core.

https://github.com/sveinnfannar/finagle/blob/a7fbf5370dffd88a0d054b9e45ca2ca4b6ee3592/CHANGES#L57..L59

kevinoliver replied:
indeed you did! apologies for the noise.

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/Endpoint.scala:
```diff
@@ -1,4 +1,4 @@
-package com.twitter.finagle.zipkin.thrift
+package com.twitter.finagle.zipkin.core
```
ditto re breaking change.

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala:
```diff
@@ -0,0 +1,149 @@
+import scala.language.reflectiveCalls
+
+/**
+ * Receives the Finagle generated traces and sends them off to Zipkin via scribe.
```
can you update the comment, as this isn't tied to scribe anymore, right?

sveinnfannar replied:
good catch 👍

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala:
```diff
@@ -0,0 +1,149 @@
+  private[this] val spanMap: DeadlineSpanMap =
+    new DeadlineSpanMap(sendSpans, 120.seconds, statsReceiver, timer)
+
+  protected[core] def flush() = spanMap.flush()
```
since you are modifying this file, if you don't mind making the world ever so slightly better, can you annotate this with the return type?

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala:
```diff
@@ -0,0 +1,149 @@
+   * Mutate the Span with whatever new info we have.
+   * If we see an "end" annotation we remove the span and send it off.
+   */
+  protected def mutate(traceId: TraceId)(f: MutableSpan => Unit) {
```
ditto re return type annotation

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala:
```diff
@@ -0,0 +1,149 @@
+  private[this] val TrueBB: ByteBuffer = ByteBuffer.wrap(Array[Byte](1))
+  private[this] val FalseBB: ByteBuffer = ByteBuffer.wrap(Array[Byte](0))
+
+  def record(record: Record) {
```
ditto re return type annotation

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala:
```diff
@@ -0,0 +1,149 @@
+   * Sets the endpoint in the span for any future annotations. Also
+   * sets the endpoint in any previous annotations that lack one.
+   */
+  protected def setEndpoint(record: Record, ia: InetSocketAddress) {
```
ditto re return type annotation

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala:
```diff
@@ -0,0 +1,149 @@
+    key: String,
+    value: ByteBuffer,
+    annotationType: thrift.AnnotationType
+  ) {
```
ditto re return type annotation

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/RawZipkinTracer.scala:
```diff
@@ -0,0 +1,149 @@
+  /**
+   * Add this record as a time based annotation.
+   */
+  protected def annotate(record: Record, value: String) {
```
ditto re return type annotation

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/Sampler.scala:
```diff
@@ -1,4 +1,4 @@
-package com.twitter.finagle.zipkin.thrift
+package com.twitter.finagle.zipkin.core
```
ditto re breaking change.

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/SamplingTracer.scala:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+
+object SamplingTracer {
```
ditto re breaking change.

kevinoliver replied:
and likewise for the other public classes in this file

sveinnfannar replied:
I replaced `c.t.f.zipkin.thrift.SamplingTracer` with a `SamplingTracer` that extends `c.t.f.zipkin.core.SamplingTracer` with a default constructor. This is because the [service loader](https://github.com/twitter/finagle/blob/develop/finagle-zipkin/src/main/resources/META-INF/services/com.twitter.finagle.tracing.Tracer) loads `c.t.f.zipkin.thrift.SamplingTracer`.

Now I realize I can replace `c.t.f.zipkin.thrift.SamplingTracer` with `c.t.f.zipkin.thrift.ScribeZipkinTracer` and load that with the service loader.
I'm not familiar with the intricacies service loading, do you think that will break anything?

Actions speak louder than words so maybe this commit explains this better than I do: https://github.com/twitter/finagle/pull/513/commits/262614b8e5a6d3d005aeb62aebece1565234ce6f

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/Span.scala:
```diff
@@ -1,4 +1,4 @@
-package com.twitter.finagle.zipkin.thrift
+package com.twitter.finagle.zipkin.core
```
ditto re breaking change.

kevinoliver commented on project/Build.scala:
```diff
@@ -304,16 +305,26 @@ object Finagle extends Build {
+      sharedSettings
+  ).settings(
+    name := "finagle-zipkin-core",
+    libraryDependencies ++= Seq(util("codec"), util("events")) ++ scroogeLibs,
```
the pants build file also has a dep on util-app and util-core. util-core should def be added here. if this doesn't have a direct util-app dependency, then can you remove from the pants build file? (or vice-versa add it here).

sveinnfannar replied:
Gotta run, I'll fix this tomorrow 👍

kevinoliver commented on finagle-zipkin-core/src/main/scala/BUILD:
```diff
@@ -0,0 +1,21 @@
+scala_library(name='scala',
```
thanks for doing all the pants build files! above and beyond what's exepcted! 💯

kevinoliver commented:
@sveinnfannar this is great. my comments are all superficial and should be easy to patch up.

sveinnfannar commented:
@kevinoliver thanks for a thorough code review!

kevinoliver commented:
my pleasure. let me know when its ready for another look.

sveinnfannar commented:
@kevinoliver I addressed the comments, can you take another look?
The changes are all pretty superficial besides this one: https://github.com/twitter/finagle/pull/513/commits/262614b8e5a6d3d005aeb62aebece1565234ce6f

taylorleese commented on finagle-zipkin/src/main/scala/com/twitter/finagle/zipkin/thrift/ScribeRawZipkinTracer.scala:
```diff
@@ -0,0 +1,184 @@
+    val transport = ClientBuilder()
+      .name("zipkin-tracer")
+      .hosts(new InetSocketAddress(scribeHost, scribePort))
+      .codec(ThriftClientFramedCodec())
```
@kevinoliver @mosesn Should we move this to the stack equivalent?

kevinoliver replied:
it should, but its not appropriate for this commit.

taylorleese commented on finagle-zipkin/src/main/scala/com/twitter/finagle/zipkin/thrift/ScribeRawZipkinTracer.scala:
```diff
@@ -0,0 +1,184 @@
+    private[this] val transport = new TReusableMemoryTransport(baos)
+    val protocol = Protocols.binaryFactory().getProtocol(transport)
+
+    def reset() = transport.reset()
```
FFTI, but could add a return type here.

taylorleese commented on finagle-zipkin/src/main/scala/com/twitter/finagle/zipkin/thrift/ScribeRawZipkinTracer.scala:
```diff
@@ -0,0 +1,184 @@
+  * Makes sure we don't trace the Scribe logging.
+  */
+private class TracelessFilter[Req, Rep] extends SimpleFilter[Req, Rep] {
+  def apply(request: Req, service: Service[Req, Rep]) = {
```
FFTI but could add a return type.

kevinoliver commented on CHANGES:
```diff
@@ -56,6 +57,14 @@ Breaking API Changes
+    `BinaryAnnotation`, `Endpoint`, `Sampler` and `SamplingTracer` to finagle-zipkin-core.
+    ``RB_ID=??``
+
+  * Builds are now only for Java 8 and Scala 2.11. See the
```
this moved up, you should remove this one.

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/SamplingTracer.scala:
```diff
@@ -24,14 +23,14 @@ private object Json {
-      @JsonDeserialize(contentAs=classOf[java.lang.Long]) traceId: Option[Long],
-      @JsonDeserialize(contentAs=classOf[java.lang.Long]) spanId: Option[Long],
-      data: Annotation)
+                       id: String,
```
lets leave the indentation as it was for the constructor params

kevinoliver commented on finagle-zipkin-core/src/main/scala/com/twitter/finagle/zipkin/core/SamplingTracer.scala:
```diff
@@ -110,82 +107,32 @@ object ZipkinTracer {
-    sink: Sink)
-  extends Tracer
-{
+  underlyingTracer: Tracer,
```
ditto re formatting

kevinoliver commented:
@sveinnfannar ok looking good to me. mind fixing this round and pinging when its good?

mosesn commented:
LGTM except for @kevinoliver's fixes

sveinnfannar commented:
Hi @kevinoliver, I addressed your comments.
And subsequently changed to emacs since IntelliJ kept formatting random code segments like a crazy person. Sorry for all the formatting diffs.

sveinnfannar commented:
@adriancole, I'm putting the finishing touches on the kafka PR, stay tuned :)

kevinoliver commented:
cool @sveinnfannar looks good now. i'll work on getting this merged in. thanks.

kevinoliver commented:
hmm @sveinnfannar i'm running into some conflicts applying the commits locally. do you mind rebasing on `develop` and then squashing these down to a single commit?

sveinnfannar commented:
Yes no problem

sveinnfannar commented:
@kevinoliver done 👍

kevinoliver commented:
ok, this one has decent sized work on our end to get it integrated. i'll keep at it but it may take a few days.

sveinnfannar commented:
ok, just ping me on gitter if I can do anything to help

kevinoliver commented:
@sveinnfannar i made good progress on this today and should get it merged in by the end of the week.

kevinoliver commented:
alrighty! got this merged in locally and should show up on the `develop` branch this coming monday.

thanks for contribution.

sveinnfannar commented:
Great to hear. Thanks for the help.

roanta commented:
Thanks! This was merged and landed: https://github.com/twitter/finagle/commit/989edc58084fa01b6e6557ff5b6adda594ba469f
